lib/fluent/plugin/filter_viaq_data_model.rb

```diff
@@ -225,6 +225,69 @@ def delempty(thing)
       thing
     end
 
+    # https://github.com/ViaQ/elasticsearch-templates/blob/master/namespaces/_default_.yml#L63
+    NORMAL_LEVELS = {
+      'emerg'    => 'emerg',
+      'panic'    => 'emerg',
+      'alert'    => 'alert',
+      'crit'     => 'crit',
+      'critical' => 'crit',
+      'err'      => 'err',
+      'error'    => 'err',
+      'warning'  => 'warning',
+      'warn'     => 'warning',
+      'notice'   => 'notice',
+      'info'     => 'info',
+      'debug'    => 'debug',
+      'trace'    => 'trace',
+      'unknown'  => 'unknown',
+    }
+    # numeric levels for the PRIORITY field
+    PRIORITY_LEVELS = {
+      0 => 'emerg',
+      1 => 'alert',
+      2 => 'crit',
+      3 => 'err',
+      4 => 'warning',
+      5 => 'notice',
+      6 => 'info',
+      7 => 'debug',
+      8 => 'trace',
+      9 => 'unknown',
+      '0' => 'emerg',
+      '1' => 'alert',
+      '2' => 'crit',
+      '3' => 'err',
+      '4' => 'warning',
+      '5' => 'notice',
+      '6' => 'info',
+      '7' => 'debug',
+      '8' => 'trace',
+      '9' => 'unknown',
+    }
+    def normalize_level(level, newlevel, stream=nil, priority=nil)
+      # if the record already has a level field, and it looks like one of our well
+      # known values, convert it to the canonical normalized form - otherwise,
+      # preserve the value in string format
+      retlevel = nil
+      if !level.nil?
+        unless retlevel = NORMAL_LEVELS[level]
+          unless level.respond_to?(:downcase) && (retlevel = NORMAL_LEVELS[level.downcase])
```

Comment:
We are only key'ing on string entries and I'm pretty certain all types have a 'to_s' method.  Why not just level.to_s.downcase to get a value since string always responsed to downcase.

Reply:
@jcantrill I would prefer to keep this code as-is.  Checking for `downcase` is essentially an optimization and a check to see if `level` is a String - if it isn't a string/doesn't have the `downcase` method then we can skip the NORMAL_LEVELS lookup, and go straight to the "use the value as-is" case.